### PR TITLE
[Lethicia Renissa] iP

### PR DESCRIPTION
**Pull request by lethiciars**
DukePro frees your mind of having to remember things you need to do. It's, 
* text-based
* easy to learn
* ~~FAST~~ _SUPER_ FAST to use

All you need to do is,
1. download it from [here](https://nus-cs2103-ay2122s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :wink:

And its __FREE__!

Features:
- [x] Managing Tasks
- [x] Managing deadlines (coming soon) 
- [ ] Reminders (coming soon)
***
If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

### CODE REVIEW

File: src/main/java/duke/Duke.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+            tasks = new TaskList(new ArrayList<Task>());
+        }
+    }
+
+    /**
+     * runs Duke.
+     */
+    public void run() {
+        Ui.showWelcomeMessage();
+        Scanner scan = new Scanner(System.in);
```
Comment (squeakysquak) [inline src/main/java/duke/Duke.java]: This is very good coding standard adherance. The spaces between the equals sign is very nice.

File: src/main/java/duke/DukeException.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package duke;
+
+/**
+ * An Exception class for Duke specific Exceptions.
```
Comment (squeakysquak) [inline src/main/java/duke/DukeException.java:4]: This is very good coding standard adherance. The javadoc comment is very nice.

File: src/main/java/duke/parser/Parser.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    /**
+     * Parses user input into command for execution.
+     *
+     * @param input full user input string
+     * @return the command based on user input
+     * @throws DukeException if user input string is not a valid format
+     */
+    public static Command getCommand(String input) throws DukeException {
+        if (input.trim().equals("bye")){
+            return new ExitCommand();
```
Comment (squeakysquak) [inline src/main/java/duke/parser/Parser.java:42]: This is very good coding standard adherance. The egyptian brackets and indentation is very nice.

**Review by squeakysquak — COMMENTED** [review]
This is very good coding standard adherance.

File: src/main/java/duke/TaskList.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+    /**
+     * Formats the task list as a numbered list.
+     *
+     * @return String showing all tasks
+     */
+    public String toDisplay() {
+        String str = "";
+        int i = 1;
+        for (Task item:TaskList) {
```
Comment (bklimey) [inline src/main/java/duke/TaskList.java]: Minor nit: Maybe [colon could be surrounded by white space](https://se-education.org/guides/conventions/java/intermediate.html#layout)?
`Layout` > `White space within a statement`

File: src/main/java/duke/command/ExitCommand.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    /**
+     * executes the command on the specified tasklist
+     *
+     * @param taskList tasklist to be operated on
+     * @throws DukeException
+     */
+    @Override
+    public String execute(TaskList taskList){
```
Comment (luluyousef) [inline src/main/java/duke/command/ExitCommand.java]: I think you might've missed the space before the braces, can add them to follow coding standard :)

File: src/main/java/duke/command/DeleteCommand.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+import duke.DukeException;
+import duke.tasks.Task;
+import duke.TaskList;
+import duke.ui.Ui;
+
+/**
+ * A DeleteCommand class encapsulates the instructions for Duke to delete a task.
+ */
+public class DeleteCommand extends Command {
+    private int TaskNo;
```
Comment (bklimey) [inline src/main/java/duke/command/DeleteCommand.java:12]: Maybe `TaskNo` could be in camelCase (i.e. `taskNo`)?

From [guildeline](https://se-education.org/guides/conventions/java/intermediate.html#naming):
> Variable names must be in camelCase.

File: src/main/java/duke/command/FindCommand.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+import duke.DukeException;
+import duke.TaskList;
+import duke.tasks.Task;
+import duke.tasks.taskType;
+import duke.ui.Ui;
+
+import java.util.ArrayList;
+
+public class FindCommand extends Command{
```
Comment (luluyousef) [inline src/main/java/duke/command/FindCommand.java]: I think you might've missed the space after Command too, can add them to follow coding standard :)

File: src/main/java/duke/command/ListCommand.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    /**
+     * executes the command on the specified tasklist
+     *
+     * @param taskList tasklist to be operated on
+     * @throws DukeException
+     */
+    @Override
+    public String execute(TaskList taskList){
```
Comment (luluyousef) [inline src/main/java/duke/command/ListCommand.java]: can also add space before braces here to follow coding standard :)

**Comment by luluyousef** [comment]
Overall, your code looks very readable to me and follows the coding conventions well! Maybe try to fix the spacing before braces in your code to follow the coding conventions more accurately :) Good job!!!

File: src/main/java/duke/command/FindCommand.java (hunk 7 above)
Comment (bklimey) [inline src/main/java/duke/command/FindCommand.java]: Minor suggestion: Maybe `CommandFinder` instead of `FindCommand` since 'Find' in this case is a verb?

From [guideline on naming](https://se-education.org/guides/conventions/java/intermediate.html#naming):
> Class/enum names must be nouns and written in PascalCase.

File: src/main/java/duke/command/FindCommand.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+     * @throws DukeException
+     */
+    @Override
+    public String execute(TaskList taskList) {
+        ArrayList<Task> results = taskList.find(keyword);
+        return Ui.showFindResults(results);
+    }
+
+    /**
+     * returns the type of command
```
Comment (bklimey) [inline src/main/java/duke/command/FindCommand.java:35]: Conventionally, the description starts with a cap and ends with a fullstop (i.e. `Returns the type of command.`).

File: src/main/java/duke/TaskList.java [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+
+    /**
+     * Adds a Task to the tasklist.
+     *
+     * @param type type of task
+     * @param details description of the task
+     */
+    public void addTask(taskType type, String[] details) {
+        switch (type) {
+        case TODO:
```
Comment (bklimey) [inline src/main/java/duke/TaskList.java:33]: I like that you used a switch statement to keep the code neat! 👍

File: src/main/java/duke/parser/Parser.java [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+        } else if (input.matches("done (.*)")) {
+            try {
+                String text = input.substring(5);
+                int taskNo = Integer.parseInt(text);
+                return new DoneCommand(taskNo);
+            } catch (NumberFormatException e) {
+                String str = "Please use the format done <task No.>";
+                throw new DukeException(str);
+            }
+        } else if (input.matches("delete (.*)")) {
```
Comment (bklimey) [inline src/main/java/duke/parser/Parser.java:49]: I like your use of regular expression! 👍

File: src/main/java/duke/parser/Parser.java [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+            } catch (ArrayIndexOutOfBoundsException e) {
+                throw new DukeException("Please use format deadline <description> /by <time>.");
+            }
+        } else if (input.matches("event(.*)")) {
+            try {
+                int start = input.indexOf("event ");
+                String validInput = input.substring(start);
+                String taskDesc = validInput.substring(6);
+                String[] fields = taskDesc.split(" /at ", 2);
+                //make sure not empty
```
Comment (bklimey) [inline src/main/java/duke/parser/Parser.java]: Maybe rephrase as `Ensure that it is not empty`? Additionally, maybe it could be more descriptive (e.g. what is not empty)?

From [guildeline on comments](https://se-education.org/guides/conventions/java/intermediate.html#comments):
>All comments should be written in English.
>Furthermore, use American spelling and avoid local slang.
 >Rationale: The code is meant for an international audience.

**Review by bklimey — COMMENTED** [review]
It is a pleasure to read the code as they are neat and tidy, and there are good use of descriptive names. Generally, I did not find any major issue. Keep it up! :)
